### PR TITLE
feat(cancellation): Provide AbortSignal to queryFn

### PR DESCRIPTION
**Pull request by jcready**
All queryFn calls will now have an AbortSignal instance passed to them
as part of the QueryFunctionContext. In environments that support the
AbortController/AbortSignal the consumer may use this `signal` as a
way to get notified about the query being cancelled. Consumers no
longer need to add a `.cancel()` function to the result of `queryFn()`

```ts
const query = useQuery('todos', async ({ signal }) => {
  const res = await fetch('/todos', {
    method: 'get',
    // Pass the signal to your request
    signal,
  });

  const json = await res.json();
  let data = null;

  if (!signal.aborted) {
    // do expensive CPU transformation of data
    data = expensiveTransform(json);
  }

  return data;
});
```

The previous behavior for cancellation will continue to work. If a
`.cancel()` function is provided as the result of calling `queryFn()`
the AbortSignal will not be notified when the cancellation happens.
If the runtime environment does not support AbortController/AbortSignal
the old behavior will continue to work, but the `signal` provided to
the `queryFn()` will not be notified when the cancellation happens.

Ref https://github.com/tannerlinsley/react-query/discussions/1671#discussioncomment-1310752
Ref https://github.com/tannerlinsley/react-query/discussions/1671#discussioncomment-1453013

Fix https://github.com/tannerlinsley/react-query/issues/2775
Fix https://github.com/tannerlinsley/react-query/issues/1893
Fix https://github.com/tannerlinsley/react-query/issues/1265
Fix https://github.com/tannerlinsley/react-query/issues/886

### CODE REVIEW

**Comment by TkDodo** [comment]
can you please once more update with the latest master, so that we get bundle size changes reported? (they won't be posted on the issue, but visible in the pipeline). Thanks.

File: docs/src/pages/guides/query-cancellation.md [hunk 1]
```diff
@@ -26,50 +49,100 @@ const query = useQuery('todos', () => {
...
+  <button onClick={(e) => {
+    e.preventDefault();
+    queryClient.cancelQueries(queryKey);
+   }}>Cancel</button>
+)
+```
+
+## Old `cancel` function
+
+Don't worry! The previous cancellation functionality will continue to work. But we do recommend that you move away from [the withdrawn cancelable promise proposal](https://github.com/tc39/proposal-cancelable-promises) to the [new `AbortSignal` interface](#_top) which has been [stardardized](https://dom.spec.whatwg.org/#interface-abortcontroller) as a general purpose construct for aborting ongoing activities in [most browsers](https://caniuse.com/abortcontroller) and in [Node](https://nodejs.org/api/globals.html#globals_class_abortsignal).
```
[the remark this answers is not in the document]
Comment (TkDodo) [reply docs/src/pages/guides/query-cancellation.md:116]: Maybe something like this?

```suggestion
Don't worry! The previous cancellation functionality will continue to work. But we do recommend that you move away from [the withdrawn cancelable promise proposal](https://github.com/tc39/proposal-cancelable-promises) to the [new `AbortSignal` interface](#_top) which has been [stardardized](https://dom.spec.whatwg.org/#interface-abortcontroller) as a general purpose construct for aborting ongoing activities in [most browsers](https://caniuse.com/abortcontroller) and in [Node](https://nodejs.org/api/globals.html#globals_class_abortsignal). The old cancel function might be removed in a future major version.
```

**Review by TkDodo — APPROVED** [review]

**Comment by jcready** [comment]
@TkDodo merged

**Comment by TkDodo** [comment]
Thank you for this great contribution @jcready . I've announced it on twitter: https://twitter.com/TkDodo/status/1454521789938294787

please let me know if you have a twitter handle so I can add you to the shoutout: https://twitter.com/TkDodo/status/1454521795529355265

**Comment by jcready** [comment]
Thank you @TkDodo! It was a pleasure working with you on this. I don't have Twitter, but I appreciate the link to my GH.

**Comment by caub** [comment]
I don't really understand, the `useQuery(key, cb)` 's cb receive a `signal`, but how do you trigger cancel? is there a `query.cancel()` function added to the API?

**Comment by jcready** [comment]
@caub https://react-query.tanstack.com/guides/query-cancellation#manual-cancellation

**Comment by TkDodo** [comment]
@caub queries will be cancelled automatically by RQ in some situations, like when the query goes to inactive state (=you’re not using it anymore) and there is a fetch pending. For manual cancellation, see the docs link above

**Comment by caub** [comment]
Ah ok thanks guys, so in all cases, whether we plan to cancel queries manually or not, it's better to pass down `signal` right? as react-query can make use of it as you just said (component unmounted, etc..)

**Comment by jcready** [comment]
@caub that is debatable :) though personally I will be using the signal in every query function.

One tradeoff is that for any query that is cancelled where either a `.cancel()` function is provided on the returned promise OR the `signal` from the query context is consumed:

- The query cache will not be populated with the resolution of your query function even if it would've been successful. Perhaps this is obvious, but it's an important difference from the old/normal behavior (no `.cancel()` function provided and `signal` not consumed).
- The query will have its state reverted back to whatever it was before the query ran https://github.com/tannerlinsley/react-query/blob/16233b0eac8a67da1b5b95e981a94b31addbfce1/src/core/tests/query.test.tsx#L241-L275

**Comment by telekstr** [comment]
@TkDodo It might be helpful to mention the version number after which this feature is available in the docs.  I was getting type errors with the destructured signal argument, and it took a bit of digging to realize this is a brand new feature (I was on 3.27.0). I would submit a PR with the update, but am not sure I have enough context to get the details right.  Just a thought.  Thanks for the great feature, @jcready -- very cool.

**Comment by TkDodo** [comment]
@telekstr good idea. The version where we introduced it is 3.30.0: https://github.com/tannerlinsley/react-query/releases/tag/v3.30.0

Please contribute it to the [Query cancellation](https://react-query.tanstack.com/guides/query-cancellation) docs page, thank you.

**Comment by telekstr** [comment]
> @telekstr good idea. The version where we introduced it is 3.30.0: https://github.com/tannerlinsley/react-query/releases/tag/v3.30.0
> 
> Please contribute it to the [Query cancellation](https://react-query.tanstack.com/guides/query-cancellation) docs page, thank you.

Sure thing, here you go: https://github.com/tannerlinsley/react-query/pull/2887

**Pull request merged**